File: 백준/Gold/1707. 이분 그래프/이분 그래프.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;
import java.util.ArrayList;
import java.util.List;
import java.util.ArrayDeque;
import java.util.Queue;
import java.util.Arrays;

public class Main {
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        // 테스트 수
        int K = Integer.parseInt(st.nextToken());
        for (int t = 0; t < K; t++) {
            
            // 정점, 간선의 개수
            st = new StringTokenizer(br.readLine());
            int v = Integer.parseInt(st.nextToken());
            int e = Integer.parseInt(st.nextToken());
            
            // 그래프 초기화
            List<List<Integer>> graph = new ArrayList<>(v + 1);
            for (int i = 0; i < v + 1; i++) {
                graph.add(new ArrayList<>());
            }
            
            // 간선 정보 입력
            for (int i = 0; i < e; i++) {
                st = new StringTokenizer(br.readLine());
                int a = Integer.parseInt(st.nextToken());
                int b = Integer.parseInt(st.nextToken());
                graph.get(a).add(b);
                graph.get(b).add(a);
            }
            
            // 이분 그래프 판별 결과
            if (isBipartite(graph, v)) {
                System.out.println("YES");
            } else {
                System.out.println("NO");
            }
        }
    }
    
    static boolean isBipartite(List<List<Integer>> graph, int N) {
        int[] colors = new int[N + 1];
        Arrays.fill(colors, 0);
        
        for (int start = 1; start <= N; start++) {
            if (colors[start] != 0) continue;
            
            Queue<Integer> q = new ArrayDeque<>();
            q.add(start);
            colors[start] = 1;
        
            while (!q.isEmpty()) {
                int node = q.poll();
                
                for (int neighbor: graph.get(node)) {
                    if (colors[neighbor] == 0) {
                        q.add(neighbor);
                        colors[neighbor] = -colors[node];
                        
                    } else if (colors[neighbor] == colors[node]) {
                        return false;
                    }
                }
            }
        }
        return true;
    }
}

File: 백준/Gold/1707. 이분 그래프/이분 그래프.py
from collections import deque
import sys
input = sys.stdin.readline

def isBipartite(graph, n):
    colors = [0] * (n + 1)  # 0 if not visited else 1 or -1
    
    # 비연결 그래프 고려 모든 노드 탐색
    for start in range(1, n + 1):
        if colors[start] != 0: 
            continue  # 이미 방문한 경우 무시
            
        # 방문 처리
        q = deque([start])
        colors[start] = 1
        
        while q:
            current_node = q.popleft()
            for neighbor in graph[current_node]:
                # 방문하지 않은 인접 노드는 다른 색으로 칠함
                if colors[neighbor] == 0:
                    q.append(neighbor)
                    colors[neighbor] = -colors[current_node]
                # 인접 노드와 색이 같다면 이분 그래프가 아님
                elif colors[neighbor] == colors[current_node]:
                    return False
    return True
    
    
# 테스트 횟수
K = int(input())
for _ in range(K):
    
    # 정점, 간선의 개수
    v, e = map(int, input().split())
    
    # 그래프 초기화
    graph = [[] for _ in range(v + 1)]
    for _ in range(e):
        a, b = map(int, input().split())
        graph[a].append(b)
        graph[b].append(a)
        
    # 이분 그래프 판별 
    print("YES" if isBipartite(graph, v) else "NO")
    